File: src/lib/offline-coach.ts
import { parseMealText } from "./foods";
import type { AppState, CoachAction } from "./types";
import { dayIntake, trendWeights, type Targets } from "./science";
import { todayKey } from "./store";
import { n } from "./format";

/**
 * مدرب محلي يعمل بلا إنترنت أو بلا مفتاح API. قواعد مبنية على نفس
 * مبادئ المدرب الذكي، حتى لا يبقى المستخدم بدون سند أبداً.
 */
export function offlineReply(text: string, s: AppState, t: Targets): { reply: string; actions: CoachAction[] } {
  const q = text.replace(/[أإآ]/g, "ا").replace(/ة/g, "ه");
  const today = s.days[todayKey()];
  const eaten = today ? dayIntake(today) : 0;
  const proteinIn = today ? today.meals.reduce((a, m) => a + m.protein, 0) : 0;
  const left = t.kcal - eaten;

  const water = q.match(/(\d+|كوب|كوبين|قلاص|ماي|ماء|مويه)/);
  if (/(شربت|اشرب).*(ماء|ماي|مويه)/.test(q) && water) {
    const n = /كوبين/.test(q) ? 2 : Number(q.match(/\d+/)?.[0] ?? 1);
    return { reply: `ممتاز 💧 سجّلها وكمّل. الماء قبل الوجبة يساعد على الشبع.`, actions: [{ type: "log_water", cups: Math.min(10, n) }] };
  }

  const items = parseMealText(text);
  if (items.length) {
    const actions: CoachAction[] = items.map(({ food, qty }) => ({
      type: "log_meal",
      name: qty === 1 ? food.name : `${food.name} ×${n(qty)}`,
      kcal: Math.round(food.kcal * qty),
      protein: Math.round(food.protein * qty),
    }));
    const kcal = actions.reduce((a, x) => a + (x.type === "log_meal" ? x.kcal : 0), 0);
    const prot = actions.reduce((a, x) => a + (x.type === "log_meal" ? x.protein : 0), 0);
    const remaining = left - kcal;
    const proteinGap = t.protein - proteinIn - prot;
    const tip =
      proteinGap > 30
        ? `باقي عليك حوالي ${n(proteinGap)} غ بروتين؛ خل وجبتك الجاية تبدأ ببروتين (زبادي يوناني، بيض، تونة).`
        : remaining < 0
          ? "طلعت فوق هدف اليوم شوي — عادي جداً. لا تعوّض بالحرمان؛ بكرة نرجع للخطة، والليلة مشي ١٠ دقائق بعد الأكل."
          : `يبقى لك تقريباً ${n(remaining)} سعرة اليوم. 👌`;
    return {
      reply: `حسبتها تقريبياً: ${n(kcal)} سعرة و${n(prot)} غ بروتين. ${tip}`,
      actions,
    };
  }

  if (/(تعبان|تعبانه|مرهق|ما عندي طاقه|كسلان|ما لي خلق|ماني قادر)/.test(q)) {
    return {
      reply: "أفهمك، والتعب مو فشل. اليوم نبي أصغر خطوة تحافظ على الخيط: دقيقتين حركة وأنت جالس، وبعدها أنت حر.",
      actions: [{ type: "start_workout", routineId: "reset-2" }],
    };
  }

  if (/(الليل|بالليل|اخر الليل|سهر|قبل النوم).*(جوع|اكل|جوعان|اشتهي)|(جوع|جوعان|اشتهي).*(الليل|بالليل)/.test(q)) {
    return {
      reply: "جوع الليل غالباً تعب أو ملل أكثر منه جوع حقيقي. جرّب ٥ دقائق تهدئة، وإذا بعدها جوعان: زبادي يوناني أو شاي بدون سكر.",
      actions: [
        { type: "start_workout", routineId: "night-5" },
        { type: "add_if_then", when: "إذا جاني جوع بعد الساعة ٩", then: "أشرب كوب ماء أو شاي، وإذا استمر آكل زبادي يوناني" },
      ],
    };
  }

  if (/(عزيمه|عزومه|وليمه|مناسبه|عرس|ضيوف|مطعم)/.test(q)) {
    return {
      reply: "العزايم جزء من حياتنا، ما نبي نهرب منها. الخطة: صحن واحد، نصه سلطة ومشاوي، ربع رز، وتحلية صغيرة إذا تبي. وقبلها بساعتين وجبة بروتين خفيفة عشان ما توصل جوعان.",
      actions: [{ type: "add_if_then", when: "إذا عندي عزيمة", then: "آكل بروتين خفيف قبلها وآخذ صحن واحد: نص خضار، ربع بروتين، ربع رز" }],
    };
  }

  if (/(ما نزل|زاد وزني|وزني زاد|ثابت|ما تغير|نفس الوزن)/.test(q)) {
    const days = Object.values(s.days).filter((d) => d.weightKg !== null).map((d) => ({ date: d.date, kg: d.weightKg as number }));
    const tr = trendWeights(days);
    const trendLine =
      tr.length >= 2 ? ` خط اتجاهك: ${n(tr[0].trend)} ← ${n(tr[tr.length - 1].trend)} كغ.` : "";
    return {
      reply: `الميزان اليومي يتأثر بالماء والملح والنوم، ممكن يتحرك كيلو بيوم واحد. نحكم على الاتجاه الأسبوعي مو القراءة.${trendLine} استمر أسبوعين بتسجيل صادق، وسند يعدّل هدفك تلقائياً حسب حرقك الحقيقي.`,
      actions: [],
    };
  }

  if (/(حلا|حلويات|سكر|شوكولا|ابي حلو)/.test(q)) {
    return {
      reply: "ما في أكل ممنوع. خذ حصة صغيرة وأنت جالس ومستمتع، بعد وجبة فيها بروتين — مو على جوع. والأهم: سجّلها بدون تأنيب.",
      actions: [],
    };
  }

  const energyNote = today?.energy === 1 ? "بما إن طاقتك اليوم تحت، " : "";
  return {
    reply: `${energyNote}خلنا نركّز على شي واحد: ${proteinIn < t.protein / 2 ? `البروتين (أكلت ${n(proteinIn)} من ${n(t.protein)} غ)` : "حركة قصيرة بعد وجبتك الجاية"}. قل لي وش أكلت اليوم وأحسبه لك، أو قل "تعبان" وأعطيك أخف خطة.`,
    actions: [],
  };
}

export function coachContext(s: AppState, t: Targets): string {
  const p = s.profile!;
  const today = s.days[todayKey()];
  const eaten = today ? dayIntake(today) : 0;
  const prot = today ? today.meals.reduce((a, m) => a + m.protein, 0) : 0;
  const weights = Object.values(s.days)
    .filter((d) => d.weightKg !== null)
    .map((d) => ({ date: d.date, kg: d.weightKg as number }));
  const tr = trendWeights(weights);
  const lastTrend = tr.length ? tr[tr.length - 1].trend : p.startWeightKg;
  return [
    `name: ${p.name}; sex: ${p.sex}; age: ${p.age}; height_cm: ${p.heightCm}`,
    `start_kg: ${p.startWeightKg}; trend_kg: ${lastTrend}; goal_kg: ${p.goalWeightKg}`,
    `targets: ${t.kcal} kcal, ${t.protein} g protein, ${t.steps} steps, ${t.water} cups water`,
    `today: eaten ${eaten} kcal, protein ${prot} g, water ${today?.water ?? 0} cups, energy ${today?.energy ?? "not checked in"} (1 low–3 high), free minutes ${today?.time ?? "?"}`,
    `today_meals: ${today?.meals.map((m) => `${m.name} ${m.kcal}kcal`).join(", ") || "none"}`,
    `why: ${p.why || "-"}; barriers: ${p.barriers.join(",") || "-"}; ramadan_mode: ${p.ramadan}`,
    `health_flags: ${p.flags.join(",") || "none"}`,
    `if_then_plans: ${p.ifThens.map((r) => `${r.when} → ${r.then}`).join(" | ") || "none"}`,
    `local_time: ${new Date().toLocaleTimeString("en-GB", { hour: "2-digit", minute: "2-digit" })}`,
  ].join("\n");
}

File: android/core/src/main/kotlin/app/sanad/core/OfflineCoach.kt
package app.sanad.core

import kotlin.math.roundToInt

data class CoachReply(val text: String, val actions: List<CoachAction>)

/**
 * مدرب محلي يعمل بلا إنترنت أو بلا مفتاح API — نفس مبادئ المدرب الذكي
 * بقواعد بسيطة، حتى لا يبقى المستخدم بدون سند.
 */
fun offlineReply(text: String, state: AppState, t: Targets, today: String): CoachReply {
    val q = normalizeArabic(text)
    val day = state.days[today]
    val proteinIn = day?.protein ?: 0
    val left = t.kcal - (day?.intake ?: 0)

    if (Regex("(شربت|اشرب).*(ماء|ماي|مويه)").containsMatchIn(q)) {
        val n = if ("كوبين" in q) 2 else Regex("\\d+").find(q)?.value?.toIntOrNull() ?: 1
        return CoachReply("ممتاز 💧 سجّلها وكمّل. الماء قبل الوجبة يساعد على الشبع.", listOf(CoachAction.LogWater(n.coerceIn(1, 10))))
    }

    val items = parseMealText(text)
    if (items.isNotEmpty()) {
        val actions = items.map { (food, qty) ->
            CoachAction.LogMeal(if (qty == 1.0) food.name else "${food.name} ×${ar(qty)}", (food.kcal * qty).roundToInt(), (food.protein * qty).roundToInt())
        }
        val kcal = actions.sumOf { it.kcal }
        val prot = actions.sumOf { it.protein }
        val gap = t.protein - proteinIn - prot
        val remaining = left - kcal
        val tip = when {
            gap > 30 -> "باقي عليك حوالي ${ar(gap)} غ بروتين؛ خل وجبتك الجاية تبدأ ببروتين (زبادي يوناني، بيض، تونة)."
            remaining < 0 -> "طلعت فوق هدف اليوم شوي — عادي جداً. لا تعوّض بالحرمان؛ بكرة نرجع للخطة، والليلة مشي ١٠ دقائق بعد الأكل."
            else -> "يبقى لك تقريباً ${ar(remaining)} سعرة اليوم. 👌"
        }
        return CoachReply("حسبتها تقريبياً: ${ar(kcal)} سعرة و${ar(prot)} غ بروتين. $tip", actions)
    }

    if (Regex("(تعبان|تعبانه|مرهق|ما عندي طاقه|كسلان|ما لي خلق|ماني قادر)").containsMatchIn(q)) {
        return CoachReply(
            "أفهمك، والتعب مو فشل. اليوم نبي أصغر خطوة تحافظ على السلسلة: دقيقتين حركة وأنت جالس، وبعدها أنت حر.",
            listOf(CoachAction.StartWorkout("reset-2")),
        )
    }
    if (Regex("(الليل|بالليل|قبل النوم|سهر).*(جوع|اكل|جوعان|اشتهي)|(جوع|جوعان|اشتهي).*(الليل|بالليل)").containsMatchIn(q)) {
        return CoachReply(
            "جوع الليل غالباً تعب أو ملل أكثر منه جوع حقيقي. جرّب ٥ دقائق تهدئة، وإذا بعدها جوعان: زبادي يوناني أو شاي بدون سكر.",
            listOf(
                CoachAction.StartWorkout("night-5"),
                CoachAction.AddIfThen("إذا جاني جوع بعد الساعة ٩", "أشرب كوب ماء أو شاي، وإذا استمر آكل زبادي يوناني"),
            ),
        )
    }
    if (Regex("(عزيمه|عزومه|وليمه|مناسبه|عرس|ضيوف|مطعم)").containsMatchIn(q)) {
        return CoachReply(
            "العزايم جزء من حياتنا، ما نبي نهرب منها. الخطة: صحن واحد، نصه سلطة ومشاوي، ربع رز، وتحلية صغيرة إذا تبي. وقبلها بساعتين وجبة بروتين خفيفة.",
            listOf(CoachAction.AddIfThen("إذا عندي عزيمة", "آكل بروتين خفيف قبلها وآخذ صحن واحد: نص خضار، ربع بروتين، ربع رز")),
        )
    }
    if (Regex("(ما نزل|زاد وزني|وزني زاد|ثابت|ما تغير|نفس الوزن)").containsMatchIn(q)) {
        val tr = trendWeights(weightPoints(state.days.values))
        val line = if (tr.size >= 2) " خط اتجاهك: ${ar(tr.first().trend)} ← ${ar(tr.last().trend)} كغ." else ""
        return CoachReply(
            "الميزان اليومي يتأثر بالماء والملح والنوم، ممكن يتحرك كيلو بيوم. نحكم على الاتجاه الأسبوعي مو القراءة.$line استمر أسبوعين بتسجيل صادق، وسند يعدّل هدفك تلقائياً.",
            emptyList(),
        )
    }
    val profile = state.profile
    if (profile != null && (Regex("(رمضان|صيام|صايم|سحور)").containsMatchIn(q) || (profile.ramadan && Regex("(فطور|افطار)").containsMatchIn(q)))) {
        val plan = ramadanPlan(profile, t)
        val split = plan.meals.joinToString("\n") { "${it.name}: ${ar(it.kcal)} سعرة و${ar(it.protein)} غ بروتين" }
        val lead = if (profile.ramadan) "خطتك الرمضانية لليوم:" else "إذا صايم، هذا توزيع هدفك على رمضان (فعّل وضع رمضان من صفحة التقدّم):"
        return CoachReply("$lead\n$split\n${plan.meals.first().tip}\n${plan.move}", emptyList())
    }
    if (profile != null && Regex("(اسبوعي|الاسبوع|مراجعه|تقييمي|كيف ماشي)").containsMatchIn(q)) {
        val r = weeklyReview(profile, state.days, t, today)
        return CoachReply("مراجعة آخر ٧ أيام: حضرت ${ar(r.activeDays)} من ٧، وتمرنت ${ar(r.workouts)} مرات.\n${r.win}\n${r.focusText}", emptyList())
    }
    if (Regex("(حلا|حلويات|سكر|شوكولا|ابي حلو)").containsMatchIn(q)) {
        return CoachReply("ما في أكل ممنوع. خذ حصة صغيرة وأنت مستمتع، بعد وجبة فيها بروتين — مو على جوع. وسجّلها بدون تأنيب.", emptyList())
    }
    val focus = if (proteinIn < t.protein / 2) "البروتين (أكلت ${ar(proteinIn)} من ${ar(t.protein)} غ)" else "حركة قصيرة بعد وجبتك الجاية"
    val energyNote = if (day?.energy == Energy.LOW) "بما إن طاقتك اليوم تحت، " else ""
    return CoachReply("${energyNote}خلنا نركّز على شي واحد: $focus. قل لي وش أكلت وأحسبه لك، أو قل \"تعبان\" وأعطيك أخف خطة.", emptyList())
}
